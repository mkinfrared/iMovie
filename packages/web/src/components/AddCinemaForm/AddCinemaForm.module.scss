@import "src/styles/common";

.AddCinemaForm {
  margin: 0 auto;
  max-width: 480px;
}
