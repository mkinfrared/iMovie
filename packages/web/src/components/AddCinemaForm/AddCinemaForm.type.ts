export interface AddCinemaFormProps {
  open: boolean;
  onClose: () => void;
}

export interface Country {
  alpha2Code: string;
  name: string;
  alpha3Code: string;
  callingCode: number;
  region: string;
  subregion: string;
  nativeName: string;
  flag: string;
  cioc: string;
  createdAt: string;
  updatedAt: string;
}

export interface State {
  id: number;
  name: string;
  abbreviation: string;
  countryId: string;
  country: Country;
  createdAt: string;
  updatedAt: string;
}

export interface City {
  id: number;
  name: string;
  zipcode: [Zipcode];
  stateId: number;
  state: State;
  countryId: string;
  country: Country;
  createdAt: string;
  updatedAt: string;
}

export interface Zipcode {
  id: number;
  code: string;
  longitude: string;
  latitude: string;
  cityId: number;
  countryId: string;
  stateId: number;
  city: City;
  country: Country;
  state: State;
  createdAt: string;
  updatedAt: string;
}
