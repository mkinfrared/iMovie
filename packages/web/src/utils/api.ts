import axios, { AxiosError } from "axios";
import { OptionsObject, SnackbarKey, SnackbarMessage } from "notistack";

import NotificationAction from "ui/NotificationAction";
import { BASE_URL } from "utils/secrets";

type ShowSnackbar = (
  message: SnackbarMessage,
  options?: OptionsObject
) => SnackbarKey;

const api = axios.create({
  baseURL: BASE_URL,
  withCredentials: true
});

const configureInterceptors = (showSnackbar: ShowSnackbar) => {
  api.interceptors.response.use(
    (config) => config,
    (error: AxiosError) => {
      if (error.message === "Network Error") {
        showSnackbar("Network Error", {
          variant: "error",
          action: NotificationAction
        });

        return;
      }

      return Promise.reject(error);
    }
  );
};

export default api;

export { configureInterceptors };
