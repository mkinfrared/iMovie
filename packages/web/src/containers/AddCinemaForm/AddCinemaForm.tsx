import React from "react";
import { useDispatch } from "react-redux";

import AddCinemaFormFC from "components/AddCinemaForm";
import { AddCinemaFormProps } from "components/AddCinemaForm/AddCinemaForm.type";

const AddCinemaForm: React.FC<Omit<AddCinemaFormProps, "dispatch">> = (
  props
) => {
  const dispatch = useDispatch();

  return <AddCinemaFormFC {...props} dispatch={dispatch} />;
};

export default React.memo(AddCinemaForm);
