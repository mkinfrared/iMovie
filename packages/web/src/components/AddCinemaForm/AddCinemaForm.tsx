import Button from "@material-ui/core/Button";
import Dialog from "@material-ui/core/Dialog";
import DialogActions from "@material-ui/core/DialogActions";
import DialogContent from "@material-ui/core/DialogContent";
import DialogContentText from "@material-ui/core/DialogContentText";
import DialogTitle from "@material-ui/core/DialogTitle";
import TextField from "@material-ui/core/TextField";
import React, { useCallback, useEffect, useState } from "react";
import { useForm } from "react-hook-form";

import { fetchCinemas } from "store/reducers/cinemas/actions";
import { Country, Zipcode } from "store/reducers/cinemas/types";
import SearchField from "ui/SearchField";
import Transition from "ui/Transition/Transition";
import api from "utils/api";
import { addCinemaFormValidation } from "utils/validation";

import css from "./AddCinemaForm.module.scss";
import { AddCinemaFormProps } from "./AddCinemaForm.type";

type FormData = {
  cinemaName: string;
  country: Pick<Country, "alpha2Code" | "name">;
  zipcode: string;
};

const AddCinemaForm = ({ open, onClose, dispatch }: AddCinemaFormProps) => {
  const [countries, setCountries] = useState<Country[]>([]);
  const [countriesLoading, setCountriesLoading] = useState(false);
  const [zip, setZip] = useState<Zipcode | null>(null);

  const {
    register,
    getValues,
    setValue,
    errors,
    triggerValidation,
    handleSubmit,
    formState
  } = useForm<FormData>({
    validationSchema: addCinemaFormValidation,
    mode: "onBlur"
  });

  const fetchCountries = async () => {
    try {
      setCountriesLoading(true);

      const response = await api.get("/country");

      setCountries(response.data.result);
    } catch (e) {
    } finally {
      setCountriesLoading(false);
    }
  };

  const fetchZipcodes = useCallback(
    async (zipcodeValue: string) => {
      const { country } = getValues({ nest: true });
      const { alpha2Code } = country;

      try {
        const response = await api.get(
          `/zipcode/${alpha2Code}/${zipcodeValue}`
        );

        setZip(response.data);
      } catch (e) {}
    },
    [getValues]
  );

  const handleZipcodeBlur = useCallback(() => {
    const { zipcode } = getValues({ nest: true });

    fetchZipcodes(zipcode);
  }, [fetchZipcodes, getValues]);

  const onSubmit = handleSubmit(async (values) => {
    try {
      const data = {
        name: values.cinemaName,
        zipcodeId: zip?.id
      };

      await api.post("/cinema", data);

      dispatch(fetchCinemas());

      onClose();
    } catch (e) {}
  });

  const handleCountryChange = useCallback(
    (_: any, country: Country | null) => {
      if (country) {
        setValue("country", country);
      }
    },
    [setValue]
  );

  const handleCountryBlur = useCallback(() => {
    triggerValidation();
  }, [triggerValidation]);

  const getOptionSelected = useCallback(
    (option, value) => option.alpha2Code === value.alpha2Code,
    []
  );

  useEffect(() => {
    fetchCountries();
  }, []);

  useEffect(() => {
    register({ name: "country" });

    register({ name: "zipcode" });
  }, [register]);

  const zipcodeDisabled = !getValues({ nest: true }).country?.alpha2Code;

  return (
    <Dialog
      open={open}
      className={css.AddCinemaForm}
      TransitionComponent={Transition}
      onClose={onClose}
      fullWidth
    >
      <DialogTitle>Add Cinema</DialogTitle>
      <DialogContent>
        <form onSubmit={onSubmit} className={css.form} data-testid="form">
          <SearchField
            options={countries}
            onChange={handleCountryChange}
            onBlur={handleCountryBlur}
            loading={countriesLoading}
            helperText={errors.country?.name?.message || null}
            error={!!errors.country?.name?.message}
            className={css.input}
            name="country"
            label="Country"
            getOptionLabel={(option) => option.name}
            getOptionSelected={getOptionSelected}
          />
          <TextField
            fullWidth
            disabled={zipcodeDisabled}
            name="zipcode"
            className={css.input}
            label="Zipcode"
            variant="outlined"
            onBlur={handleZipcodeBlur}
            helperText={errors.zipcode?.message || null}
            error={!!errors.zipcode}
            inputRef={register}
            inputProps={{ "data-testid": "zipcode" }}
          />
          <TextField
            fullWidth
            name="cinemaName"
            className={css.input}
            label="Cinema Name"
            variant="outlined"
            helperText={errors.cinemaName?.message || null}
            error={!!errors.cinemaName}
            inputRef={register}
            inputProps={{ "data-testid": "cinemaName" }}
          />
        </form>
      </DialogContent>
      <DialogContent>
        <DialogContentText variant="body1">
          Country: {getValues({ nest: true }).country?.name}
        </DialogContentText>
        <DialogContentText variant="body1">
          State: {zip?.state.name}
        </DialogContentText>
        <DialogContentText variant="body1">
          City: {zip?.city.name}
        </DialogContentText>
        <DialogContentText variant="body1">
          Cinema: {getValues().cinemaName}
        </DialogContentText>
      </DialogContent>
      <DialogActions>
        <Button onClick={onClose}>Cancel</Button>
        <Button
          data-testid="loginButton"
          onClick={onSubmit}
          color="primary"
          disabled={formState.isSubmitting}
        >
          Add
        </Button>
      </DialogActions>
    </Dialog>
  );
};

export default React.memo(AddCinemaForm);
