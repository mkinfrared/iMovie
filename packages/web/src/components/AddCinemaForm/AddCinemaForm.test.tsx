import { act, findByTestId, render } from "@testing-library/react";
import React, { ReactElement } from "react";

import api from "utils/api";

import AddCinemaForm from "./AddCinemaForm";

jest.mock("utils/api");

jest.mock("ui/SearchField", () => (props: any) => (
  <div data-testid="ui/SearchField">
    <code>{JSON.stringify(props)}</code>
  </div>
));

describe("<AddCinemaForm />", () => {
  const apiMock = api as jest.Mocked<typeof api>;
  const open = true;
  const onClose = jest.fn();
  const dispatch = jest.fn();

  let Component: ReactElement;

  beforeEach(() => {
    Component = (
      <AddCinemaForm onClose={onClose} open={open} dispatch={dispatch} />
    );
  });

  it("should be defined", async () => {
    apiMock.get.mockReturnValueOnce(Promise.resolve({ data: [42] }));

    await act(async () => {
      const { baseElement } = await render(Component);

      await findByTestId(baseElement, "ui/SearchField");
    });
  });

  it("should match a snapshot", () => {
    const { baseElement } = render(Component);

    expect(baseElement).toMatchSnapshot();
  });
});
