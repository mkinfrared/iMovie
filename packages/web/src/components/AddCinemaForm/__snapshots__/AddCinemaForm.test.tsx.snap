// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddCinemaForm /> should match a snapshot 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root AddCinemaForm"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper"
      role="none presentation"
      style="webkit-transform: none; transform: none; webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullWidth MuiPaper-elevation24 MuiPaper-rounded"
        role="dialog"
      >
        <div
          class="MuiDialogTitle-root"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6"
          >
            Add Cinema
          </h2>
        </div>
        <div
          class="MuiDialogContent-root"
        >
          <form
            class="form"
            data-testid="form"
          >
            <div
              data-testid="ui/SearchField"
            >
              <code>
                {"options":[],"loading":true,"helperText":null,"error":false,"className":"input","name":"country","label":"Country"}
              </code>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root input MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-disabled Mui-disabled"
                data-shrink="false"
              >
                Zipcode
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root Mui-disabled Mui-disabled MuiInputBase-fullWidth MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled Mui-disabled"
                  data-testid="zipcode"
                  disabled=""
                  name="zipcode"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-322 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-324"
                  >
                    <span>
                      Zipcode
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root input MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                data-shrink="false"
              >
                Cinema Name
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input"
                  data-testid="cinemaName"
                  name="cinemaName"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-322 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-324"
                  >
                    <span>
                      Cinema Name
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </form>
        </div>
        <div
          class="MuiDialogContent-root"
        >
          <p
            class="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            Country: 
          </p>
          <p
            class="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            State: 
          </p>
          <p
            class="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            City: 
          </p>
          <p
            class="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            Cinema: 
          </p>
        </div>
        <div
          class="MuiDialogActions-root MuiDialogActions-spacing"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Cancel
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
            data-testid="loginButton"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Add
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
